units/en/unit2/smolagents/code_agents.mdx

```diff
@@ -0,0 +1,378 @@
+
+![Code vs JSON Actions](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/transformers/code_vs_json_actions.png)
+Graphic from the paper [Executable Code Actions Elicit Better LLM Agents](https://huggingface.co/papers/2402.01030)
+
+<Tip>
+If you want to learn more about why code agents are effective, check out <a href="https://huggingface.co/docs/smolagents/en/conceptual_guides/intro_agents#code-agents" target="_blank">this guide</a> from the smolagents documentation. 
+</Tip>
+
+## Why Code Agents?
+
+In a multi-step agent process, the LLM writes and executes actions, typically involving external tool calls. Traditional approaches use a JSON format to specify tool names and arguments as strings, **which the system must parse to determine which tool to execute**.
+
+However, research shows that **tool-calling LLMs work more effectively with code directly**. This is a core principle of `smolagents`, as shown in the diagram above from [Executable Code Actions Elicit Better LLM Agents](https://huggingface.co/papers/2402.01030).
+
+Writing actions in code rather than JSON offers several key advantages:
+
+* **Composability**: Easily combine and reuse actions
+* **Object Management**: Work directly with complex structures like images
+* **Generality**: Express any computationally possible task
+* **Natural for LLMs**: High-quality code is already present in LLM training data
+
+## How Does a Code Agent Work?
+
+![From https://huggingface.co/docs/smolagents/conceptual_guides/react](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/smolagents/codeagent_docs.png)
+
+The diagram above illustrates how `CodeAgent.run()` operates, following the [ReAct framework](https://huggingface.co/papers/2210.03629), which is currently the preferred approach for building multi-step agents. We already introduced the `MultiStepAgent` in the previous section, which serves as the core building block for `smolagents`. The `CodeAgent` is a special kind of `MultiStepAgent`, as we will see in this example underneath.
+
+A `CodeAgent` performs actions through a cycle of steps, with existing variables and knowledge being incorporated into the agent's logs as follows:
+
+1. The system prompt is stored in a `SystemPromptStep`, and the user query is logged in a `TaskStep`.
+
+2. Then, the following while loop is executed:
+
+    2.1 The method `agent.write_memory_to_messages()` writes the agent's logs into a list of LLM-readable [chat messages](https://huggingface.co/docs/transformers/en/chat_templating).
+    
+    2.2 These messages are sent to a `Model`, which generates a completion. 
+    
+    2.3 The completion is parsed to extract the action, which, in our case, could be a code snippet since we're working with a `CodeAgent`.
```

Comment:
```suggestion
    2.3 The completion is parsed to extract the action, which, in our case, should be a code snippet since we're working with a `CodeAgent`.
```